src/modules/siptrace/siptrace.c

```diff
@@ -636,85 +690,93 @@ static int sip_trace_store_db(siptrace_data_t *sto)
 {
-	char *duri;
-	sip_uri_t uri;
-	dest_info_t *dst = NULL;
-	proxy_l_t *p = NULL;
-	str dup_uri_str = {0, 0};
+	int idx;
+	enum siptrace_type_t trace_type = SIPTRACE_NONE;
 
-	if(param_no != 1) {
-		LM_DBG("params:%s\n", (char *)*param);
-		return 0;
+	if (sflags == NULL || sflags->s == NULL || sflags->len == 0) {
+		return SIPTRACE_NONE;
 	}
 
-	if(((char *)(*param))[0] == '\0') {
-		// Empty URI, use the URI set at module level (dup_uri)
-		if(dup_uri) {
-			uri = *dup_uri;
-		} else {
-			LM_ERR("Missing duplicate URI\n");
-			return -1;
-		}
-	} else {
-		duri = (char *)*param;
+	for (idx = 0; idx < sflags->len; idx++) {
+		switch(sflags->s[idx]|0x20) { /*|0x20 - to lowercase */
+			case SIPTRACE_MESSAGE:
+			case SIPTRACE_TRANSACTION:
+			case SIPTRACE_DIALOG:
+				if (trace_type != SIPTRACE_NONE) {
+					LM_ERR("only one tracing flag can be used <%.*s>!\n",
+							sflags->len, sflags->s);
+					return SIPTRACE_NONE;
+				}
 
-		if(!(*duri)) {
-			LM_ERR("invalid dup URI\n");
-			return -1;
+				trace_type = (sflags->s[idx]|0x20);
+				break;
+			case ' ':
+			case '\t':
+				break;
+			default:
+				LM_ERR("Invalid character <%c> in <%.*s> at position <%d>!\n", sflags->s[idx],
+						sflags->len, sflags->s, idx);
+				return SIPTRACE_NONE;
 		}
+	}
 
-		LM_DBG("sip_trace URI:%s\n", duri);
+	return trace_type;
+}
 
-		dup_uri_str.s = duri;
-		dup_uri_str.len = strlen(dup_uri_str.s);
-		memset(&uri, 0, sizeof(struct sip_uri));
+static int fixup_siptrace(void **param, int param_no)
+{
+	str sflags;
+	enum siptrace_type_t trace_type;
 
-		if(parse_uri(dup_uri_str.s, dup_uri_str.len, &uri) < 0) {
-			LM_ERR("bad dup uri\n");
-			return -1;
-		}
+	if(param_no < 1 || param_no > 3) {
+		LM_DBG("params:%s\n", (char *)*param);
+		return 0;
 	}
 
-	dst = (dest_info_t *)pkg_malloc(sizeof(dest_info_t));
-	if(dst == 0) {
-		LM_ERR("no more pkg memory left\n");
-		return -1;
-	}
-	init_dest_info(dst);
-	/* create a temporary proxy*/
-	dst->proto = PROTO_UDP;
-	p = mk_proxy(&uri.host, (uri.port_no) ? uri.port_no : SIP_PORT, dst->proto);
-	if(p == 0) {
-		LM_ERR("bad host name in uri\n");
-		pkg_free(dst);
-		return -1;
-	}
-	hostent2su(&dst->to, &p->host, p->addr_idx, (p->port) ? p->port : SIP_PORT);
+	if (param_no == 1 || param_no == 2) {
+		/* correlation id */
+		return fixup_spve_all(param, param_no);
+	} else if (param_no == 3) {
+		/* tracing type; string only */
+		sflags.s = (char *)*param;
+		sflags.len = strlen(sflags.s);
 
-	pkg_free(*param);
-	/* free temporary proxy*/
-	if(p) {
-		free_proxy(p); /* frees only p content, not p itself */
-		pkg_free(p);
+		trace_type = parse_siptrace_flag(&sflags);
+		if (trace_type == SIPTRACE_NONE) {
+			LM_ERR("Failed to parse trace type!\n");
+			return -1;
+		}
+
+		*param = pkg_malloc(sizeof(trace_type));
+		memcpy(*param, &trace_type, sizeof(trace_type));
 	}
 
-	*param = (void *)dst;
 	return 0;
 }
 
+
 /**
- * Send sip trace with destination and correlation id
+ * TODO TODO TODO:
```

Comment:
Is this still a TODO or was already added?

Reply:
No, thanks.